### PR TITLE
Add ContactResultMap class to reduce heap allocations for multiple contact requests

### PR DESCRIPTION
**Pull request by Levi-Armstrong**
Initial testing shows this can have a significant performance bump. See results bellow where the last two columns show % reduction in Time, CPU Time. In case with large number of contacts the reduction is around 99%.

image

### CODE REVIEW

**Comment by Levi-Armstrong** [comment]
Baseline heaptrack summary for reference. I will post the results after with updates.
heaptrack_baseline_planner_unit

**Comment by Levi-Armstrong** [comment]
@marip8 If you have time it would not hurt to have another set of eyes on this PR before it gets merged.

**Comment by Levi-Armstrong** [comment]
Just finished adding the unit tests

**Comment by Levi-Armstrong** [comment]
Improvements results. This change has a significant impact on Peak Memory, Heap Allocation and Performance. 

Highlights

- TrajOpt
  - Performance
    - Simple 88% decrease
    - Complex 3% decrease
  - Peak Memory: ~30MB (25%) decrease
  - Memory Allocations: 9 Million (35%) decrease
- TrajOpt Ifopt
  - Performance
    - Simple 89% decrease
    - Complex 4% decrease
  - Peak Memory: ~33MB (79%) decrease
  - Memory Allocations: 3 Million (38%) decrease

TrajOpt
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<html>
<body>


  |   |   |   |   | Test Body |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
ID | Simple Unit(us) | % Change | Planning Unit (ms) | % Change | Peak Memory | % Change | Allocations | Temp Allocations | Allocation Sum | % Change | Description
Baseline | 2,860.00 |   | 23,432.00 |   | 111.40 |   | 24,739,478.00 | 885,722.00 | 25,625,200.00 |   |  
1 | 332.00 | -88.39% | 22,741.00 | -2.95% | 83.00 | -25.49% | 16,175,975.00 | 470,917.00 | 16,646,892.00 | -35.04% | Sel


</body>

</html>

TrajOpt Ifopt:

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<html>

<body>


  |   |   |   |   | Test Body |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
ID | Simple Unit(us) | % Change | Planning Unit (ms) | % Change | Peak Memory | % Change | Allocations | Temp Allocations | Allocation Sum | % Change | Description
Baseline | 2,786.00 |   | 1,041.00 |   | 41.60 |   | 9,170,119.00 | 262,537.00 | 9,432,656.00 |   |  
1 | 303.00 | -89.12% | 999.00 | -4.03% | 8.70 | -79.09% | 5,738,612.00 | 24,028.00 | 5,762,640.00 | -38.91% |  


</body>

</html>

**Comment by Levi-Armstrong** [comment]
Improvements results including the last two commits

Highlights

- TrajOpt
  - Performance
    - Simple 88% decrease
    - Complex 3% decrease
  - Peak Memory: ~34MB (30%) decrease
  - Memory Allocations: 14.5 Million (57%) decrease
- TrajOpt Ifopt
  - Performance
    - Simple 89% decrease
    - Complex 10% decrease
  - Peak Memory: ~34MB (81%) decrease
  - Memory Allocations: 7.3 Million (77%) decrease

TrajOpt
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<html>
<body>


  |   |   |   |   | Test Body |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
ID | Simple Unit(us) | % Change | Planning Unit (ms) | % Change | Peak Memory | % Change | Allocations | Temp Allocations | Allocation Sum | % Change | Description
Baseline | 2,860.00 |   | 23,432.00 |   | 111.40 |   | 24,739,478.00 | 885,722.00 | 25,625,200.00 |   |  
1 | 331.00 | -88.43% | 23,222.00 | -0.90% | 77.40 | -30.52% | 10,669,590.00 | 397,571.00 | 11,067,161.00 | -56.81%


</body>

</html>

TrajOpt Ifopt:

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN">

<html>

<body>


  |   |   |   |   | Test Body |   |   |   |   |   |  
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
ID | Simple Unit(us) | % Change | Planning Unit (ms) | % Change | Peak Memory | % Change | Allocations | Temp Allocations | Allocation Sum | % Change | Description
Baseline | 2,786.00 |   | 1,041.00 |   | 41.60 |   | 9,170,119.00 | 262,537.00 | 9,432,656.00 |   |  
1 | 296.00 | -89.38% | 933.00 | -10.37% | 7.80 | -81.25% | 2,098,024.00 | 21,060.00 | 2,119,084.00 | -77.53%


</body>

</html>

File: tesseract_collision/core/src/types.cpp
Comment (mpowelson) [inline tesseract_collision/core/src/types.cpp:95]: I think this can just be  `cv.reserve(results.size())` if you want

File: tesseract_collision/core/src/types.cpp
Comment (mpowelson) [inline tesseract_collision/core/src/types.cpp:212]: Do these combined with clear result in any non intuitive behavior? For example, a range based for loop will iterate over things after you call clear. That's probably fine. I was trying to come up with some way that this could internally store another iterable structure that is a view into the actual data store, but that's probably not worth it.

**Review by mpowelson — APPROVED** [review]

**Comment by Levi-Armstrong** [comment]
As a sanity check to make sure it was not an issue passing an object that contains a thread_local by reference into multiple threads is an issue. It was that it is not an issue and even though it was created in a different thread and passed by reference to  two threads it still gets its own copy.

```
#include <iostream>
#include <vector>
#include <thread>

struct TempClass
{
  TempClass(int i) : cnt_(i) {}
  void fn()
  {
    thread_local std::vector<double> data;
    std::cout << "This: " << this << std::endl;
    std::cout << "Capacity: " << data.capacity() << std::endl;
    std::cout << "Address: " << &data << std::endl;
    data.reserve(cnt_);
  };
private:
  int cnt_;
};

void threadFn(TempClass& tmp)
{
    tmp.fn();
}

int main()
{

    TempClass a(5);
    auto f = [&a]()
    {
        a.fn();
    };
    std::thread t1(f);
    std::thread t2(f);

    t1.join();
    t2.join();
}
```

Results:

```
Program returned: 0
This: 0x7fff88a8429c
Capacity: 0
Address: 0x7fa15188a6e0
This: 0x7fff88a8429c
Capacity: 0
Address: 0x7fa1510896e0
```

File: tesseract_common/include/tesseract_common/types.h
Comment (Levi-Armstrong) [reply tesseract_common/include/tesseract_common/types.h:87]: Updated

File: tesseract_collision/core/include/tesseract_collision/core/types.h
Comment (Levi-Armstrong) [reply tesseract_collision/core/include/tesseract_collision/core/types.h:163]: It should assign instead of append. I updated the documentation.

File: tesseract_collision/core/src/types.cpp
Comment (Levi-Armstrong) [reply tesseract_collision/core/src/types.cpp:70]: Renamed to `count_`

File: tesseract_collision/core/src/types.cpp
Comment (Levi-Armstrong) [reply tesseract_collision/core/src/types.cpp:72]: fixed

File: tesseract_collision/core/src/types.cpp
Comment (Levi-Armstrong) [reply tesseract_collision/core/src/types.cpp:95]: Yep, updated.

File: tesseract_collision/core/include/tesseract_collision/core/types.h
Comment (Levi-Armstrong) [reply tesseract_collision/core/include/tesseract_collision/core/types.h:135]: Yea, I think a better approach would be to leverage a object pool for `ContactResultVector` then have everything else function as a normal map and update the set and add methods to pull from the object pool if the pair does not exist. I have found some examples and will most likely implement this over the next few weeks.

Comment (Levi-Armstrong) [reply tesseract_collision/core/include/tesseract_collision/core/types.h:135]: Added documentation.

File: tesseract_collision/core/src/types.cpp
Comment (Levi-Armstrong) [reply tesseract_collision/core/src/types.cpp:212]: Yea, I think a better approach would be to leverage a object pool for ContactResultVector then have everything else function as a normal map and update the set and add methods to pull from the object pool if the pair does not exist. I have found some examples and will most likely implement this over the next few weeks.

**Pull request merged**
